File: dsa/challenges/quick_sort/quick_sort.py
def quick_sort(arr, left=0, right=None):
    if right == None:
        right = len(arr)-1
    if left < right:
        pos = partition(arr,left,right)
        quick_sort(arr, left, pos-1)
        quick_sort(arr, pos+1, right)
    return arr

def partition(arr, left, right):
    pivot = arr[right]
    low = left-1
    # iterate over left to the pivot-1
    for i in range(left,right):
        if arr[i]<= pivot:
            low = low + 1
            swap(arr, i, low)
    swap(arr, right, low+1)
    return low+1

def swap(arr, i, low):
    arr[i], arr[low] = arr[low], arr[i]
    return

if __name__ == "__main__":
    arr = [5,12,7,5,5,7]
    print(quick_sort(arr))

File: dsa/java_challenges/src/main/java/java_challenges/sorts/QuickSort.java
package java_challenges.sorts;

public class QuickSort {
    public int[] quickSort(int[] arr) {
        int left = 0;
        int right = arr.length-1;
        arr = quickSort(arr, left, right);
        return arr;
    }

    public int[] quickSort(int[] arr, int left, int right) {
        if (left<right) {
            int position = partition(arr, left, right);
            arr = quickSort(arr, left, position-1);
            arr = quickSort(arr, position+1, right);
        }
        return arr;
    }
    private int[] swap(int[] arr, int low, int i){
        int temp = arr[low];
        arr[low] = arr[i];
        arr[i]= temp;
        return arr;
    }
    private int partition (int[] arr, int left, int right){
        int pivot = arr[right];
        int low = left-1;
        for (int i = left; i<right; i++){
            if (arr[i]<=pivot){
                low ++;
                arr = swap(arr, low, i);
            }
        }
        low = low+1;
        arr = swap(arr, low, right);
        return low;
    }

}
